// @aws-amplify/geo/lib-esm/internals/InternalGeo.js
import { __awaiter, __generator, __read } from "tslib";
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { Amplify, GeoAction, ConsoleLogger as Logger, parseAWSExports, } from '@aws-amplify/core';
import { AmazonLocationServiceProvider } from '../Providers/AmazonLocationServiceProvider';
import { validateCoordinates } from '../util';
import { getGeoUserAgentDetails } from './utils';
var logger = new Logger('Geo');
var DEFAULT_PROVIDER = 'AmazonLocationService';
var InternalGeoClass = /** @class */ (function () {
    function InternalGeoClass() {
        this._config = {};
        this._pluggables = [];
        logger.debug('Geo Options', this._config);
    }
    /**
     * get the name of the module category
     * @returns {string} name of the module category
     */
    InternalGeoClass.prototype.getModuleName = function () {
        return InternalGeoClass.MODULE;
    };
    /**
     * add plugin into Geo category
     * @param {Object} pluggable - an instance of the plugin
     */
    InternalGeoClass.prototype.addPluggable = function (pluggable) {
        if (pluggable && pluggable.getCategory() === 'Geo') {
            this._pluggables.push(pluggable);
            var config = pluggable.configure(this._config[pluggable.getProviderName()]);
            return config;
        }
    };
    /**
     * Get the plugin object
     * @param providerName - the name of the plugin
     */
    InternalGeoClass.prototype.getPluggable = function (providerName) {
        var pluggable = this._pluggables.find(function (pluggable) { return pluggable.getProviderName() === providerName; });
        if (pluggable === undefined) {
            logger.debug('No plugin found with providerName', providerName);
            throw new Error('No plugin found in Geo for the provider');
        }
        else
            return pluggable;
    };
    /**
     * Remove the plugin object
     * @param providerName - the name of the plugin
     */
    InternalGeoClass.prototype.removePluggable = function (providerName) {
        this._pluggables = this._pluggables.filter(function (pluggable) { return pluggable.getProviderName() !== providerName; });
        return;
    };
    /**
     * Configure Geo
     * @param {Object} config - Configuration object for Geo
     * @return {Object} - Current configuration
     */
    InternalGeoClass.prototype.configure = function (config) {
        var _this = this;
        logger.debug('configure Geo');
        if (!config)
            return this._config;
        var amplifyConfig = parseAWSExports(config);
        this._config = Object.assign({}, this._config, amplifyConfig.Geo, config);
        this._pluggables.forEach(function (pluggable) {
            pluggable.configure(_this._config[pluggable.getProviderName()]);
        });
        if (this._pluggables.length === 0) {
            this.addPluggable(new AmazonLocationServiceProvider());
        }
        return this._config;
    };
    /**
     * Get the map resources that are currently available through the provider
     * @param {string} provider
     * @returns - Array of available map resources
     */
    InternalGeoClass.prototype.getAvailableMaps = function (provider) {
        if (provider === void 0) { provider = DEFAULT_PROVIDER; }
        var prov = this.getPluggable(provider);
        return prov.getAvailableMaps();
    };
    /**
     * Get the map resource set as default in amplify config
     * @param {string} provider
     * @returns - Map resource set as the default in amplify config
     */
    InternalGeoClass.prototype.getDefaultMap = function (provider) {
        if (provider === void 0) { provider = DEFAULT_PROVIDER; }
        var prov = this.getPluggable(provider);
        return prov.getDefaultMap();
    };
    /**
     * Search by text input with optional parameters
     * @param  {string} text - The text string that is to be searched for
     * @param  {SearchByTextOptions} options? - Optional parameters to the search
     * @param {CustomUserAgentDetails} customUserAgentDetails - Optional parameter to send user agent details
     * @returns {Promise<Place[]>} - Promise resolves to a list of Places that match search parameters
     */
    InternalGeoClass.prototype.searchByText = function (text, options, customUserAgentDetails) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, providerName, prov, error_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = (options || {}).providerName, providerName = _a === void 0 ? DEFAULT_PROVIDER : _a;
                        prov = this.getPluggable(providerName);
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, prov.searchByText(text, options, getGeoUserAgentDetails(GeoAction.SearchByText, customUserAgentDetails))];
                    case 2: return [2 /*return*/, _b.sent()];
                    case 3:
                        error_1 = _b.sent();
                        logger.debug(error_1);
                        throw error_1;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Search for search term suggestions based on input text
     * @param  {string} text - The text string that is to be search for
     * @param  {SearchByTextOptions} options? - Optional parameters to the search
     * @param {CustomUserAgentDetails} customUserAgentDetails - Optional parameter to send user agent details
     * @returns {Promise<SearchForSuggestionsResults>} - Resolves to an array of search suggestion strings
     */
    InternalGeoClass.prototype.searchForSuggestions = function (text, options, customUserAgentDetails) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, providerName, prov, error_2;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = (options || {}).providerName, providerName = _a === void 0 ? DEFAULT_PROVIDER : _a;
                        prov = this.getPluggable(providerName);
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, prov.searchForSuggestions(text, options, getGeoUserAgentDetails(GeoAction.SearchForSuggestions, customUserAgentDetails))];
                    case 2: return [2 /*return*/, _b.sent()];
                    case 3:
                        error_2 = _b.sent();
                        logger.debug(error_2);
                        throw error_2;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Search for location by unique ID
     * @param  {string} placeId - Unique ID of the location that is to be searched for
     * @param  {searchByPlaceIdOptions} options? - Optional parameters to the search
     * @param {CustomUserAgentDetails} customUserAgentDetails - Optional parameter to send user agent details
     * @returns {Promise<Place>} - Resolves to a place with the given placeId
     */
    InternalGeoClass.prototype.searchByPlaceId = function (placeId, options, customUserAgentDetails) {
        return __awaiter(this, void 0, void 0, function () {
            var providerName, prov, error_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        providerName = DEFAULT_PROVIDER;
                        prov = this.getPluggable(providerName);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, prov.searchByPlaceId(placeId, options, getGeoUserAgentDetails(GeoAction.SearchByPlaceId, customUserAgentDetails))];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3:
                        error_3 = _a.sent();
                        logger.debug(error_3);
                        throw error_3;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Reverse geocoding search via a coordinate point on the map
     * @param coordinates - Coordinates array for the search input
     * @param options - Options parameters for the search
     * @param {CustomUserAgentDetails} customUserAgentDetails - Optional parameter to send user agent details
     * @returns {Promise<Place>} - Promise that resolves to a place matching search coordinates
     */
    InternalGeoClass.prototype.searchByCoordinates = function (coordinates, options, customUserAgentDetails) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, providerName, prov, _b, lng, lat, error_4;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _a = (options || {}).providerName, providerName = _a === void 0 ? DEFAULT_PROVIDER : _a;
                        prov = this.getPluggable(providerName);
                        _b = __read(coordinates, 2), lng = _b[0], lat = _b[1];
                        _c.label = 1;
                    case 1:
                        _c.trys.push([1, 3, , 4]);
                        validateCoordinates(lng, lat);
                        return [4 /*yield*/, prov.searchByCoordinates(coordinates, options, getGeoUserAgentDetails(GeoAction.SearchByCoordinates))];
                    case 2: return [2 /*return*/, _c.sent()];
                    case 3:
                        error_4 = _c.sent();
                        logger.debug(error_4);
                        throw error_4;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Create geofences
     * @param geofences - Single or array of geofence objects to create
     * @param options? - Optional parameters for creating geofences
     * @param {CustomUserAgentDetails} customUserAgentDetails - Optional parameter to send user agent details
     * @returns {Promise<SaveGeofencesResults>} - Promise that resolves to an object with:
     *   successes: list of geofences successfully created
     *   errors: list of geofences that failed to create
     */
    InternalGeoClass.prototype.saveGeofences = function (geofences, options, customUserAgentDetails) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, providerName, prov, geofenceInputArray, error_5;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = (options || {}).providerName, providerName = _a === void 0 ? DEFAULT_PROVIDER : _a;
                        prov = this.getPluggable(providerName);
                        if (!Array.isArray(geofences)) {
                            geofenceInputArray = [geofences];
                        }
                        else {
                            geofenceInputArray = geofences;
                        }
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, prov.saveGeofences(geofenceInputArray, options, getGeoUserAgentDetails(GeoAction.SaveGeofences, customUserAgentDetails))];
                    case 2: return [2 /*return*/, _b.sent()];
                    case 3:
                        error_5 = _b.sent();
                        logger.debug(error_5);
                        throw error_5;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get a single geofence by geofenceId
     * @param geofenceId: GeofenceId - The string id of the geofence to get
     * @param options?: GeofenceOptions - Optional parameters for getting a geofence
     * @param {CustomUserAgentDetails} customUserAgentDetails - Optional parameter to send user agent details
     * @returns Promise<Geofence> - Promise that resolves to a geofence object
     */
    InternalGeoClass.prototype.getGeofence = function (geofenceId, options, customUserAgentDetails) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, providerName, prov, error_6;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = (options || {}).providerName, providerName = _a === void 0 ? DEFAULT_PROVIDER : _a;
                        prov = this.getPluggable(providerName);
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, prov.getGeofence(geofenceId, options, getGeoUserAgentDetails(GeoAction.GetGeofence, customUserAgentDetails))];
                    case 2: return [2 /*return*/, _b.sent()];
                    case 3:
                        error_6 = _b.sent();
                        logger.debug(error_6);
                        throw error_6;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * List geofences
     * @param  options?: ListGeofenceOptions
     * @param {CustomUserAgentDetails} customUserAgentDetails - Optional parameter to send user agent details
     * @returns {Promise<ListGeofencesResults>} - Promise that resolves to an object with:
     *   entries: list of geofences - 100 geofences are listed per page
     *   nextToken: token for next page of geofences
     */
    InternalGeoClass.prototype.listGeofences = function (options, customUserAgentDetails) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, providerName, prov, error_7;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = (options || {}).providerName, providerName = _a === void 0 ? DEFAULT_PROVIDER : _a;
                        prov = this.getPluggable(providerName);
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, prov.listGeofences(options, getGeoUserAgentDetails(GeoAction.ListGeofences, customUserAgentDetails))];
                    case 2: return [2 /*return*/, _b.sent()];
                    case 3:
                        error_7 = _b.sent();
                        logger.debug(error_7);
                        throw error_7;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Delete geofences
     * @param geofenceIds: string|string[]
     * @param options?: GeofenceOptions
     * @param {CustomUserAgentDetails} customUserAgentDetails - Optional parameter to send user agent details
     * @returns {Promise<DeleteGeofencesResults>} - Promise that resolves to an object with:
     *  successes: list of geofences successfully deleted
     *  errors: list of geofences that failed to delete
     */
    InternalGeoClass.prototype.deleteGeofences = function (geofenceIds, options, customUserAgentDetails) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, providerName, prov, geofenceIdsInputArray, error_8;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = (options || {}).providerName, providerName = _a === void 0 ? DEFAULT_PROVIDER : _a;
                        prov = this.getPluggable(providerName);
                        if (!Array.isArray(geofenceIds)) {
                            geofenceIdsInputArray = [geofenceIds];
                        }
                        else {
                            geofenceIdsInputArray = geofenceIds;
                        }
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, prov.deleteGeofences(geofenceIdsInputArray, options, getGeoUserAgentDetails(GeoAction.DeleteGeofences, customUserAgentDetails))];
                    case 2: return [2 /*return*/, _b.sent()];
                    case 3:
                        error_8 = _b.sent();
                        logger.debug(error_8);
                        throw error_8;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    InternalGeoClass.MODULE = 'InternalGeo';
    return InternalGeoClass;
}());
export { InternalGeoClass };
export var InternalGeo = new InternalGeoClass();
Amplify.register(InternalGeo);
